rustler_mix/lib/rustler/compiler.ex

```diff
@@ -122,13 +122,199 @@ defmodule Rustler.Compiler do
   defp make_build_mode_flag(args, :release), do: args ++ ["--release"]
   defp make_build_mode_flag(args, :debug), do: args
 
+  defp get_target_os_type(target) when is_binary(target) do
```

Comment:
Sure! For all the keys, they were obtained by `rustc --print target-list`. 
<details>
<summary>target keys</summary>

```
$ rustc --version
rustc 1.56.1 (59eed8a2a 2021-11-01)
$ rustc --print target-list
aarch64-apple-darwin
aarch64-apple-ios
aarch64-apple-ios-macabi
aarch64-apple-ios-sim
aarch64-apple-tvos
aarch64-fuchsia
aarch64-linux-android
aarch64-pc-windows-msvc
aarch64-unknown-freebsd
aarch64-unknown-hermit
aarch64-unknown-linux-gnu
aarch64-unknown-linux-gnu_ilp32
aarch64-unknown-linux-musl
aarch64-unknown-netbsd
aarch64-unknown-none
aarch64-unknown-none-softfloat
aarch64-unknown-openbsd
aarch64-unknown-redox
aarch64-unknown-uefi
aarch64-uwp-windows-msvc
aarch64-wrs-vxworks
aarch64_be-unknown-linux-gnu
aarch64_be-unknown-linux-gnu_ilp32
arm-linux-androideabi
arm-unknown-linux-gnueabi
arm-unknown-linux-gnueabihf
arm-unknown-linux-musleabi
arm-unknown-linux-musleabihf
armebv7r-none-eabi
armebv7r-none-eabihf
armv4t-unknown-linux-gnueabi
armv5te-unknown-linux-gnueabi
armv5te-unknown-linux-musleabi
armv5te-unknown-linux-uclibceabi
armv6-unknown-freebsd
armv6-unknown-netbsd-eabihf
armv7-apple-ios
armv7-linux-androideabi
armv7-unknown-freebsd
armv7-unknown-linux-gnueabi
armv7-unknown-linux-gnueabihf
armv7-unknown-linux-musleabi
armv7-unknown-linux-musleabihf
armv7-unknown-netbsd-eabihf
armv7-wrs-vxworks-eabihf
armv7a-none-eabi
armv7a-none-eabihf
armv7r-none-eabi
armv7r-none-eabihf
armv7s-apple-ios
asmjs-unknown-emscripten
avr-unknown-gnu-atmega328
bpfeb-unknown-none
bpfel-unknown-none
hexagon-unknown-linux-musl
i386-apple-ios
i586-pc-windows-msvc
i586-unknown-linux-gnu
i586-unknown-linux-musl
i686-apple-darwin
i686-linux-android
i686-pc-windows-gnu
i686-pc-windows-msvc
i686-unknown-freebsd
i686-unknown-haiku
i686-unknown-linux-gnu
i686-unknown-linux-musl
i686-unknown-netbsd
i686-unknown-openbsd
i686-unknown-uefi
i686-uwp-windows-gnu
i686-uwp-windows-msvc
i686-wrs-vxworks
mips-unknown-linux-gnu
mips-unknown-linux-musl
mips-unknown-linux-uclibc
mips64-unknown-linux-gnuabi64
mips64-unknown-linux-muslabi64
mips64el-unknown-linux-gnuabi64
mips64el-unknown-linux-muslabi64
mipsel-sony-psp
mipsel-unknown-linux-gnu
mipsel-unknown-linux-musl
mipsel-unknown-linux-uclibc
mipsel-unknown-none
mipsisa32r6-unknown-linux-gnu
mipsisa32r6el-unknown-linux-gnu
mipsisa64r6-unknown-linux-gnuabi64
mipsisa64r6el-unknown-linux-gnuabi64
msp430-none-elf
nvptx64-nvidia-cuda
powerpc-unknown-freebsd
powerpc-unknown-linux-gnu
powerpc-unknown-linux-gnuspe
powerpc-unknown-linux-musl
powerpc-unknown-netbsd
powerpc-unknown-openbsd
powerpc-wrs-vxworks
powerpc-wrs-vxworks-spe
powerpc64-unknown-freebsd
powerpc64-unknown-linux-gnu
powerpc64-unknown-linux-musl
powerpc64-wrs-vxworks
powerpc64le-unknown-freebsd
powerpc64le-unknown-linux-gnu
powerpc64le-unknown-linux-musl
riscv32gc-unknown-linux-gnu
riscv32gc-unknown-linux-musl
riscv32i-unknown-none-elf
riscv32imac-unknown-none-elf
riscv32imc-esp-espidf
riscv32imc-unknown-none-elf
riscv64gc-unknown-linux-gnu
riscv64gc-unknown-linux-musl
riscv64gc-unknown-none-elf
riscv64imac-unknown-none-elf
s390x-unknown-linux-gnu
s390x-unknown-linux-musl
sparc-unknown-linux-gnu
sparc64-unknown-linux-gnu
sparc64-unknown-netbsd
sparc64-unknown-openbsd
sparcv9-sun-solaris
thumbv4t-none-eabi
thumbv6m-none-eabi
thumbv7a-pc-windows-msvc
thumbv7a-uwp-windows-msvc
thumbv7em-none-eabi
thumbv7em-none-eabihf
thumbv7m-none-eabi
thumbv7neon-linux-androideabi
thumbv7neon-unknown-linux-gnueabihf
thumbv7neon-unknown-linux-musleabihf
thumbv8m.base-none-eabi
thumbv8m.main-none-eabi
thumbv8m.main-none-eabihf
wasm32-unknown-emscripten
wasm32-unknown-unknown
wasm32-wasi
wasm64-unknown-unknown
x86_64-apple-darwin
x86_64-apple-ios
x86_64-apple-ios-macabi
x86_64-apple-tvos
x86_64-fortanix-unknown-sgx
x86_64-fuchsia
x86_64-linux-android
x86_64-pc-solaris
x86_64-pc-windows-gnu
x86_64-pc-windows-msvc
x86_64-sun-solaris
x86_64-unknown-dragonfly
x86_64-unknown-freebsd
x86_64-unknown-haiku
x86_64-unknown-hermit
x86_64-unknown-illumos
x86_64-unknown-l4re-uclibc
x86_64-unknown-linux-gnu
x86_64-unknown-linux-gnux32
x86_64-unknown-linux-musl
x86_64-unknown-netbsd
x86_64-unknown-none-hermitkernel
x86_64-unknown-none-linuxkernel
x86_64-unknown-openbsd
x86_64-unknown-redox
x86_64-unknown-uefi
x86_64-uwp-windows-gnu
x86_64-uwp-windows-msvc
x86_64-wrs-vxworks
```

</details>

The corresponding values were handled manually based on [this doc](https://www.erlang.org/doc/man/os.html).


> ```
> type() -> {Osfamily, Osname}
>   Types
>     Osfamily = unix | win32
>     Osname = atom()
> ```
> Returns the `Osfamily` and, in some cases, the `Osname` of the current OS.
>
> On Unix, `Osname` has the same value as `uname -s` returns, but in lower case. For example, on Solaris 1 and 2, it is `sunos`.
>
> On Windows, `Osname` is `nt`.

And triples that correspond to `{:win32, :nt}` or `{:unix, :darwin}` are the ones in which we're interested because they use different extensions. As for the rest of them, no matter it is FreeBSD, NetBSD, or else, they all use `.so` as the file extension for dynamic libraries.

Another thing to note is that this is not intended to be the most accurate list for non-darwin and non-win32 targets because, for example, I really have no idea what would be the output of `uname -s` on `mipsel-sony-psp` (could be either `Linux` or `PSP`).

However, since the code that uses these values only distinguishes among `{:win32, _}`, `{:unix, :darwin}` and `{:unix, _}`. I think it perhaps is acceptable to use `{:unix, :linux}` for `mipsel-sony-psp` and targets starting with `wasm`? 

What do you think?

Reply:
> I think it perhaps is acceptable to use {:unix, :linux} for mipsel-sony-psp and targets starting with wasm?

As we default to `{:unix, :linux}` on `nil` as well, this is probably ok. I wonder if we should explicitly not handle targets for which we cannot determine the OS however. Note that the target platform also needs to support Erlang/ERTS, so some of  the targets are actually not viable?

I wonder if it would be better to pattern-match against `*-linux*`, `*-windows*` instead for this list. That would future-proof the function if Rust gains support on new platforms. What do you think?

Reply:
> I wonder if we should explicitly not handle targets for which we cannot determine the OS however.

I agree with you. Let's just handle the targets that we know the corresponding OSes.

> I wonder if it would be better to pattern-match against *-linux*, *-windows* instead for this list.

This seems to be good. And we can have the following patterns:

1. `*-linux*`, maps to `{:unix, :linux}`
2. `*-windows*`, maps to `{:win32, :nt}`
3. `*-apple*`, maps to `{:unix, :darwin}`
4. `*-freebsd*`, maps to `{:unix, :freebsd}`
5. `*-netbsd*`, maps to `{:unix, :netbsd}`
6. `*-openbsd*`, maps to `{:unix, :openbsd}`
7. `*-solaris*`, maps to `{:unix, sunos}`
8. everything else.

But shall we raise an exception if the target falls into category 8, or we default to `{:unix, :linux}` for them?

And the results, (I'll push the new commit in a moment)
<details>
<summary><code>{:unix, :linux}</code></summary>

```
aarch64-linux-android
aarch64-unknown-linux-gnu
aarch64-unknown-linux-gnu_ilp32
aarch64-unknown-linux-musl
aarch64_be-unknown-linux-gnu
aarch64_be-unknown-linux-gnu_ilp32
arm-linux-androideabi
arm-unknown-linux-gnueabi
arm-unknown-linux-gnueabihf
arm-unknown-linux-musleabi
arm-unknown-linux-musleabihf
armv4t-unknown-linux-gnueabi
armv5te-unknown-linux-gnueabi
armv5te-unknown-linux-musleabi
armv5te-unknown-linux-uclibceabi
armv7-linux-androideabi
armv7-unknown-linux-gnueabi
armv7-unknown-linux-gnueabihf
armv7-unknown-linux-musleabi
armv7-unknown-linux-musleabihf
hexagon-unknown-linux-musl
i586-unknown-linux-gnu
i586-unknown-linux-musl
i686-linux-android
i686-unknown-linux-gnu
i686-unknown-linux-musl
mips-unknown-linux-gnu
mips-unknown-linux-musl
mips-unknown-linux-uclibc
mips64-unknown-linux-gnuabi64
mips64-unknown-linux-muslabi64
mips64el-unknown-linux-gnuabi64
mips64el-unknown-linux-muslabi64
mipsel-unknown-linux-gnu
mipsel-unknown-linux-musl
mipsel-unknown-linux-uclibc
mipsisa32r6-unknown-linux-gnu
mipsisa32r6el-unknown-linux-gnu
mipsisa64r6-unknown-linux-gnuabi64
mipsisa64r6el-unknown-linux-gnuabi64
powerpc-unknown-linux-gnu
powerpc-unknown-linux-gnuspe
powerpc-unknown-linux-musl
powerpc64-unknown-linux-gnu
powerpc64-unknown-linux-musl
powerpc64le-unknown-linux-gnu
powerpc64le-unknown-linux-musl
riscv32gc-unknown-linux-gnu
riscv32gc-unknown-linux-musl
riscv64gc-unknown-linux-gnu
riscv64gc-unknown-linux-musl
s390x-unknown-linux-gnu
s390x-unknown-linux-musl
sparc-unknown-linux-gnu
sparc64-unknown-linux-gnu
thumbv7neon-linux-androideabi
thumbv7neon-unknown-linux-gnueabihf
thumbv7neon-unknown-linux-musleabihf
x86_64-linux-android
x86_64-unknown-linux-gnu
x86_64-unknown-linux-gnux32
x86_64-unknown-linux-musl
x86_64-unknown-none-linuxkernel
```
</details>

<details>
<summary><code>{:win32, :nt}</code></summary>

```
aarch64-pc-windows-msvc
aarch64-uwp-windows-msvc
i586-pc-windows-msvc
i686-pc-windows-gnu
i686-pc-windows-msvc
i686-uwp-windows-gnu
i686-uwp-windows-msvc
thumbv7a-pc-windows-msvc
thumbv7a-uwp-windows-msvc
x86_64-pc-windows-gnu
x86_64-pc-windows-msvc
x86_64-uwp-windows-gnu
x86_64-uwp-windows-msvc
```
</details>

<details>
<summary><code>{:unix, :darwin}</code></summary>

```
aarch64-apple-darwin
aarch64-apple-ios
aarch64-apple-ios-macabi
aarch64-apple-ios-sim
aarch64-apple-tvos
armv7-apple-ios
armv7s-apple-ios
i386-apple-ios
i686-apple-darwin
x86_64-apple-darwin
x86_64-apple-ios
x86_64-apple-ios-macabi
x86_64-apple-tvos
```
</details>

<details>
<summary><code>{:unix, :freebsd}</code></summary>

```
aarch64-unknown-freebsd
armv6-unknown-freebsd
armv7-unknown-freebsd
i686-unknown-freebsd
powerpc-unknown-freebsd
powerpc64-unknown-freebsd
powerpc64le-unknown-freebsd
x86_64-unknown-freebsd
```
</details>

<details>
<summary><code>{:unix, :netbsd}</code></summary>

```
aarch64-unknown-netbsd
armv6-unknown-netbsd-eabihf
armv7-unknown-netbsd-eabihf
i686-unknown-netbsd
powerpc-unknown-netbsd
sparc64-unknown-netbsd
x86_64-unknown-netbsd
```
</details>

<details>
<summary><code>{:unix, :openbsd}</code></summary>

```
aarch64-unknown-openbsd
i686-unknown-openbsd
powerpc-unknown-openbsd
sparc64-unknown-openbsd
x86_64-unknown-openbsd
```
</details>

<details>
<summary><code>{:unix, :openbsd}</code></summary>

```
sparcv9-sun-solaris
x86_64-pc-solaris
x86_64-sun-solaris
```
</details>

Reply:
I added a message along with the `:unknown_target` exception in this commit, 2dcb8af. And sorry, English is not my native language, so please suggest any better wording for that message.